### ISSUE TITLE
A few questions about Tutorial 4.

### ISSUE THREAD

**Issue by Jamiesona** [bentrevett/pytorch-seq2seq#90]
1. About the Encoder
    `
        hidden = torch.tanh(self.fc(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim = 1)))
    `
    when the model calculates hidden[-2] and hidden[-1], it uses the information of some pad tokens, 
    say, I got a sentence "\<sos\> thank you very much \<eos\> \<pad\> \<pad\> \<pad\>", the 
    three \<pad\> tokens will affect the value of hidden, does it matter?

2. About the Attention
    I build a model for English to Chinese translation, and I use Encoder outputs and the last hidden 
    state of Decoder to calculate attention scores.
    encoder_outputs of shape (batch_size, N, hid_dim), N is the length of input sequeence.
    last_hidden of shape (1, batch_size, hid_dim), it works like this:

    `
    last_hidden = last_hidden.permute((1, 0, 2))

    attn_score = (encoder_outs * last_hidden).sum(dim=2)

    attn_score = attn_score.masked_fill(mask, -1e10)

    attn_score = torch.softmax(attn_score, dim=1)

    return attn_score
    `

   and I got attentions like this, all the translation results are right, but it pays to mush attention on the end of a sentence.

![Figure_3](https://user-images.githubusercontent.com/20358807/77633977-37787580-6f8b-11ea-9e10-2e36e3fbcadc.png)

**Comment by bentrevett** [comment]
1. Because we used packed padded sequences the final hidden state for each sequence won't necessarily be the last element in the sequence. It will be taken from `outputs` by indexing into it with the provided source lengths.

An example, we have a batch size of 2 and a sequence length of 10. The first example in the batch is actually 10 tokens long, but the second is only 5 and thus has been padded. The hidden state for the second example in the batch is actually taken from `outputs[4][1]` (the fifth hidden state, the one over the final non-pad token) and not `outputs[-1][1]` (the hidden state taken from the last element in the sequence). 

```python
import torch
import torch.nn as nn

emb_dim = 50
hid_dim = 100

gru = nn.GRU(emb_dim, hid_dim)

batch_size = 2
sequence_length = 10 

x = torch.randn(sequence_length, batch_size, emb_dim)
lengths = torch.LongTensor([10, 5])

packed_x = nn.utils.rnn.pack_padded_sequence(x, lengths)

packed_outputs, hidden = gru(packed_x)

outputs, _ = nn.utils.rnn.pad_packed_sequence(packed_outputs) 

# outputs = [sequence length, batch size, hid dim]
# hidden = [1, batch size, hid dim]

# "final" hidden state for second example in batch is actually fifth hidden state in outputs
# this is because the second example in the batch is only 5 tokens long, as indicated by
# the lengths tensor

assert torch.all(torch.eq(outputs[4][1], hidden[:,1]))
```

**Comment by Jamiesona** [comment]
Thanks for your explaination, I am clear about the first question now, before seeing your tutorial, I used the sequeence length to get encoder outputs without pad tokens, so stupid ^-^

About the attention, it's just dot production between last hidden state and encoder output at each time step. For the first time step of decoder, last hidden state is actually the last output of encoder,  it's dot production value is larger, so it pays more attention to itself, am I right? I have replaced the last hidden with decoder's last output(after embedding, embed_dim == hid_dim), it seems better. but still not good enough.

I did use your attention implementation, and it works, anyway, I will try my best fo figure it out. I have seen some awsome repositories on your github, they are really helpful. I'm new to NLP, would you please give me some suggestions on how to learn it. In Chinese it means "大佬 求带", again thank you for your time.

**Comment by Jamiesona** [comment]
I replaced Multi30k with my own dataset, trained the model 300 epochs. The results are all correct, but still the attention score images are not that perfect. 

I notice that the Multi30k dataset contains sentences like "somebody is doing something", say "A man is is smiling at a stuffed lion", "A old man having beer alone"... , they are all descriptions of pictures.  But in my own dataset there seems no relationship between two sentences, maybe that's a reason. You can see the results here 

https://i.loli.net/2020/03/27/Jo21a4QU5PutKeI.jpg

https://i.loli.net/2020/03/27/NJzkoiOq2E8IuPg.jpg

https://i.loli.net/2020/03/27/6XSd7tqZ4ubwyNE.png

**Comment by Jamiesona** [comment]
I think I hava found why the attention does not work, it is because I used too little training data. My GPU is weak, so I can not use too much training data. Now I make use of all the training data, and I get a reasonable result just like this.

![attn2](https://user-images.githubusercontent.com/20358807/77812969-0bbad400-70e0-11ea-834a-7ae731ed1de8.png)
![attn1](https://user-images.githubusercontent.com/20358807/77812981-18d7c300-70e0-11ea-932f-2c4e5e80e50f.png)

Your tutorials really help me to learn NLP, thank you very much.
